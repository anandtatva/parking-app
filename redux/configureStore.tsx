import { persistReducer, persistStore } from "redux-persist";
import { combineReducers, createStore } from "redux";
import parkingSlotSlice from "./reducer/parking-slot-slice";
import AsyncStorage from "@react-native-async-storage/async-storage";


const persistConfig = {
    key: "parking-slot",
    storage: AsyncStorage
}

const rootReducers = combineReducers({
    parkingSlot: parkingSlotSlice
})

const slotPersistReducer = persistReducer(persistConfig, rootReducers);

const store = createStore(slotPersistReducer);

const persistor = persistStore(store)

export { store, persistor }