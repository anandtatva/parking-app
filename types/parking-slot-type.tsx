type ParkingSlotType = {
    id: string;
    slotNumber: number;
    isEmpty: boolean;
    carNumber: string;
    enterTime: null | number;
    exitTime: null | number;
}

export type {
    ParkingSlotType
}