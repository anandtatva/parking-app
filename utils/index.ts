export * from './strings';
export * from './colors';
export * from "./utils";
export * from "./alerts";