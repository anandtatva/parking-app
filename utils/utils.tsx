import moment from "moment"
function validateCarNumber(carNumber: string) {
    const pattern = /[A-Z][A-Z] [0-9][0-9] [A-Z][A-Z] [0-9][0-9][0-9][0-9]$/i;
    const re = new RegExp(pattern);
    return re.test(carNumber)
}
function msToTime(time: number) {
    return moment(time * 1000).format("DD/MM/YYYY HH:mm:ss");

}
function getDifferenceDateString(startTime: number) {
    const { hour, minutes } = getDifferenceDate(startTime, true)
    if (hour && minutes)
        return `${hour} hours ${minutes} minutes`
    else if (hour) return `${hour} hour`
    else if (minutes) return `${minutes} minutes`
    else return `${hour} hours`
}
function getDifferenceDate(enterTime: number, status = false) {
    const totalMinutes = moment().diff(moment(enterTime * 1000), "minutes")
    if (!isNaN(totalMinutes) && totalMinutes) {
        const hours = Math.floor(totalMinutes / 60)
        const minutes = totalMinutes % 60;
        if (status)
            return { hour: hours, minutes }
        else
            return { hour: hours + (minutes > 0 && 1 || 0), minutes }
    } else return { hour: 1, minutes: 0 }

}
export {
    validateCarNumber,
    msToTime,
    getDifferenceDateString,
    getDifferenceDate
}