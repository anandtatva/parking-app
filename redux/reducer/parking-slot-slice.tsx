import { createSlice } from "@reduxjs/toolkit";
import { ParkingSlotType } from "../../types/parking-slot-type";

interface StateProps {
    parkingSlots: ParkingSlotType[];
}
const initialState: StateProps = {
    parkingSlots: []
}

const parkingSlotSlice = createSlice({
    name: "parking-slot",
    initialState,
    reducers: {
        setParkingSlot: (state, action) => {
            state.parkingSlots = action.payload;
        }
    }
})

export const { setParkingSlot } = parkingSlotSlice.actions;

export default parkingSlotSlice.reducer;