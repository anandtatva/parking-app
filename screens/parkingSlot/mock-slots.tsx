import { ParkingSlotType } from "../../types/parking-slot-type";

const mockSlots: ParkingSlotType[] = [
    {
        id: '55bfe613-f4b1-4bbf-8ebf-f8a661b7519c',
        slotNumber: 1,
        isEmpty: true,
        carNumber: '',
        enterTime: null,
        exitTime: null
    },
    {
        id: '81075ab6-823e-44b3-8015-d034df652950',
        slotNumber: 2,
        isEmpty: true,
        carNumber: '',
        enterTime: null,
        exitTime: null
    }
]

const fullMockSlot: ParkingSlotType[] = [
    {
        id: '55bfe613-f4b1-4bbf-8ebf-f8a661b7519c',
        slotNumber: 1,
        isEmpty: false,
        carNumber: 'XZ 08 XZ 0000',
        enterTime: 1656554945,
        exitTime: null
    },
    {
        id: '81075ab6-823e-44b3-8015-d034df652950',
        slotNumber: 2,
        isEmpty: false,
        carNumber: 'XZ 08 XZ 0000',
        enterTime: 1656564945,
        exitTime: null
    }
]

export { mockSlots, fullMockSlot };