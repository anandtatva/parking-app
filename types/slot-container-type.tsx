import { ParkingSlotType } from "./parking-slot-type";

export interface SlotContainerType {
    item: ParkingSlotType,
    index: number
    onRemoveCar: (index: number, item: ParkingSlotType) => void
}