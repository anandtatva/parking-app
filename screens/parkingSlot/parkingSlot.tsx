import { StatusBar } from 'expo-status-bar';
import { Platform, Alert, Text, StyleSheet, View, FlatList } from 'react-native';

import EditScreenInfo from '../../components/EditScreenInfo';
import { useSelector, useDispatch } from "react-redux";
import { SlotContainer } from './slot-container';
import { ParkingSlotType } from '../../types/parking-slot-type';
import { string } from '../../utils';
import moment from "moment";
import { setParkingSlot } from '../../redux/reducer/parking-slot-slice';

export default function ParkingSlot({ navigation }) {
  const dispatch = useDispatch()
  const parkingSlots = useSelector(state => state.parkingSlot)
  const removeCarAlert = (index: number, item: ParkingSlotType) => {
    if (Platform.OS === "web") {
      const resp = confirm("Are you sure you want to remove car from parking?");
      resp && removeCar(index, item);

    } else {
      Alert.alert('', 'Are you sure you want to remove car from parking?', [
        { text: 'Cancel', onPress: () => null },
        { text: 'OK', onPress: () => removeCar(index, item) },
      ], { cancelable: false })
    }

  }

  const removeCar = (index: number, item: ParkingSlotType) => {
    let updateCar: ParkingSlotType = { ...parkingSlots.parkingSlots[index] };
    updateCar.exitTime = moment().unix()
    const tempList = JSON.parse(JSON.stringify(parkingSlots.parkingSlots));
    tempList.splice(index, 1, updateCar)
    dispatch(setParkingSlot(tempList))
    navigation.navigate("PaymentScreen", { slotDetail: updateCar })
  }
  const renderItem = ({ item, index }) => (
    <SlotContainer
      item={item}
      index={index}
      onRemoveCar={removeCarAlert}
    />
  )
  return (
    <View style={styles.container}>
      {parkingSlots?.parkingSlots?.length ? (
        <Text style={styles.helpTxt}>{string.noteRemoveCar}</Text>
      ) : null}

      <FlatList
        testID="Slot-List"
        numColumns={2}
        keyExtractor={(_, index) => String(index)}
        data={parkingSlots?.parkingSlots || []}
        renderItem={renderItem}
        ListEmptyComponent={() => (
          <Text style={styles.title} testID={"Empty-Slots"}>{string.emptySlot}</Text>
        )}
      />
    </View>
  );
}

const styles = StyleSheet.create({
  helpTxt: {
    opacity: 0.8,
    padding: 10,
    textAlign: "center"
  },
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
  },
  title: {
    fontSize: 20,
    fontWeight: 'bold',
  },
  separator: {
    marginVertical: 30,
    height: 1,
    width: '80%',
  },
});
