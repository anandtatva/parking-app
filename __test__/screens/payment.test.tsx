import { render, waitFor, act, fireEvent, cleanup } from "@testing-library/react-native";
import Payment from "../../screens/payment/payment";
import { Provider } from "react-redux";
import { store } from "../../redux/configureStore";
import { setParkingSlot } from "../../redux/reducer/parking-slot-slice";
import { mockSlots, fullMockSlot } from "../../screens/parkingSlot/mock-slots";
import * as Utils from "../../utils/utils";
import { Alert } from "react-native";
import { string } from "../../utils";
import { ParkingSlotType } from "../../types/parking-slot-type";
import moment from "moment";

jest.mock("@react-navigation/native", () => {
    const actualNav = jest.requireActual("@react-navigation/native");
    return {
        ...actualNav,
        useFocusEffect: () => jest.fn(),
        useNavigation: () => ({
            navigate: jest.fn()
        }),
        useRoute: () => ({
            params: {}
        })
    }

})
const spyAlert = jest.spyOn(Alert, 'alert');

describe("View Parking Slot", () => {
    afterEach(() => {
        jest.clearAllMocks();
        cleanup()
    })
    beforeEach(() => {
        store.dispatch(setParkingSlot([]))
    })
    it("renders correctly", () => {
        const slotDetail: ParkingSlotType = { ...fullMockSlot[0], exitTime: moment().unix() }

        const navigation = {
            navigate: jest.fn()
        }
        const tree = render(
            <Provider store={store}>
                <Payment route={{ params: slotDetail }} navigation={navigation} />
            </Provider>
        ).toJSON();
        expect(tree).toMatchSnapshot();
    })
    it("Amount should be valid", async () => {
        const slotDetail: ParkingSlotType = {
            ...fullMockSlot[0],
            exitTime: moment().unix(),
            enterTime: moment().subtract(315, "minutes").unix(),
        }
        store.dispatch(setParkingSlot([]))
        const navigation = {
            navigate: jest.fn()
        }
        const { getByTestId } = render(
            <Provider store={store}>
                <Payment route={{ params: { slotDetail } }} navigation={navigation} />
            </Provider>
        )
        const totalAmount = getByTestId("Total-Amount");
        const totalTime = getByTestId("Total-Time");
        expect(totalAmount.children[0]).toBe("$50")
        expect(totalTime.children[0]).toBe("5 hours 15 minutes")
    })
    it("payment should be done", async () => {
        const slotDetail: ParkingSlotType = {
            ...fullMockSlot[0],
            exitTime: moment().unix(),
            enterTime: moment().subtract(315, "minutes").unix(),
        }
        store.dispatch(setParkingSlot([]))
        const navigation = {
            navigate: jest.fn(),
            reset: jest.fn()
        }
        const { getByTestId } = render(
            <Provider store={store}>
                <Payment route={{ params: { slotDetail } }} navigation={navigation} />
            </Provider>
        )
        const totalAmount = getByTestId("Total-Amount");
        const totalTime = getByTestId("Total-Time");
        expect(totalAmount.children[0]).toBe("$50")
        expect(totalTime.children[0]).toBe("5 hours 15 minutes")
        const payButton = getByTestId("Pay-Button");
        fireEvent.press(payButton)
        expect(navigation.reset).toHaveBeenCalledWith({ index: 0, routes: [{ name: "CreateSlotScreen" }] })

    })

})