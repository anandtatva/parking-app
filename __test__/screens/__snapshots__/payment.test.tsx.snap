// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View Parking Slot renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
        "width": "100%",
      },
      Object {
        "padding": 20,
      },
    ]
  }
>
  <View
    style={
      Object {
        "flexDirection": "row",
        "marginTop": 10,
      }
    }
  >
    <Text
      style={
        Object {
          "fontWeight": "bold",
        }
      }
    >
      In time: 
    </Text>
    <Text>
      Invalid date
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
          "marginTop": 10,
        },
      ]
    }
  >
    <Text
      style={
        Object {
          "fontWeight": "bold",
        }
      }
    >
      Out time: 
    </Text>
    <Text>
      Invalid date
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
          "marginTop": 10,
        },
      ]
    }
  >
    <Text
      style={
        Object {
          "fontWeight": "bold",
        }
      }
    >
      Difference 
    </Text>
    <Text
      testID="Total-Time"
    >
      1 hour
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "borderColor": "black",
          "borderWidth": 1,
          "marginTop": 10,
          "padding": 10,
        },
      ]
    }
  >
    <Text
      style={
        Object {
          "fontWeight": "bold",
        }
      }
    >
      Charge Notes 
    </Text>
    <Text>
      1. First 2 hour $10
    </Text>
    <Text>
      2. $10 will be charged for every consecutive hour then after
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
          "marginTop": 10,
        },
      ]
    }
  >
    <Text
      style={
        Object {
          "fontWeight": "bold",
        }
      }
    >
      Total Amount 
    </Text>
    <Text
      testID="Total-Amount"
    >
      $10
    </Text>
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "#000",
        "opacity": 1,
        "paddingHorizontal": 8,
        "paddingVertical": 10,
      }
    }
    testID="Pay-Button"
  >
    <Text
      style={
        Object {
          "color": "#fff",
          "textAlign": "center",
        }
      }
    >
      Pay
    </Text>
  </View>
</View>
`;
